_posts/2019-09-06-totw-147.md

```diff
@@ -0,0 +1,236 @@
+repository*, then the breakage won't surface until the engineers on our project
+try to update to a newer version of that code. If we expect that those engineers
+will be willing and able to fix the `switch` statement, then all is well.
+
+However, if `AnEnum` is owned by a different project in the *same repository*
+the situation is a bit more precarious. A change to `AnEnum` might cause our
+code to greak at head, and the engineer making the change might not be willing
+or able to fix it for us. Indeed, if there were many similar exhaustive `switch`
+statements over `AnEnum`, it'd be extremely challenging for them to fix all such
+usages.
+
+For these reasons, it's best to use exhaustive `switch` statements only on
+`enum` types that either we own, or whose owner has explicitly guaranteed that
+no new enumerators will be added.
+
+In our example, let's suppose that `AnEnum` is owned by a different project, but
+the documentation promises that no new enumerators will be added. Let's add a
+comment so that future readers understand our reasoning.
+
+```c++ {.good}
+std::string AnEnumToString(AnEnum an_enum) {
+  switch (an_enum) {
+    case kFoo:
+      return "kFoo";
+    case kBar:
+      return "kBar";
+    case kBaz:
+      return "kBaz";
+    // No default. The API of AnEnum guarantees no new enumerators will be
+    // added.
+  }
+  LOG(ERROR) << "Unexpected value for AnEnum: " << an_enum;
+  return kUnknownAnEnumString;
+}
+```
+
+## Conclusions
+
+Exhaustive `switch` statements can be an excellent tool for ensuring that all
+enumerators are explicitly handled, provided that we:
+
+*   Explicitly handle the case where the `enum` has a non-enumerator value,
+    falling through the entire `switch` statement. In particular if the
+    enclosing function has a return value, we must ensure that the function
+    still either returns a value or crashes in a well-defined and debuggable
+    way.
+*   Ensure that one of:
+    *   The owner of the `enum` type either guarantees no new enumerators will
+        be added,
+    *   The owner of the `enum` is willing and able to fix our code when new
+        enumerators are aded,
+    *   If our code uses exhaustive switch statements and is broken due to an
+        enumerator being added, the owner of the `enum` is not blocked by this
+        breakage.
+
+When making an `enum` type available to other projects, we should either:
+
+*   Explicitly guarantee that no new enumerators will be added, so that users
+    can take advantage of exhaustive `switch` statements.
+*   Explicitly reserve the right to add new enumerators without notice, to
+    discourage consumers from writing exhaustive `switch` statements. One
+    idiomatic way of doing so is to add a sentinel enumerator clearly not meant
+    to be used in API consumers' exhaustive switch statements; e.g.
+    `kNotForUseWithExhaustiveSwitchStatements`.
+
+## FAQ
+
+*   Why does the compiler allow omitting a `return` statement after an
```

Comment:
Markdown is terrible for bulleted lists that have anything but the most basic format. (Additionally, our CSS causes bullets followed by whitespace anywhere in the list to show up alone on their own line. This is a bug bu alas.)

I don't think these should really be list items anyway. The bulleted parts can be replaced with bold text, IMO. (Enclose in ** ... **

Reply:
Done.